pkg/termcorder/termcorder.go

```diff
@@ -0,0 +1,210 @@
+//new variable of type `Termcording`.
+func TermcordingFromFlags(options ...func(*Termcording) error) (*Termcording, error) {
+	var fName, cmdName string
+	var cmdArgs []string
+
+	var h, q, a, k bool
+
+	cli = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	cli.BoolVar(&h, "h", false, "Prints this message")
+	cli.BoolVar(&q, "q", false, "Quiet mode - suppresses the recording start and end prompts")
+	cli.BoolVar(&a, "a", false, "Appends to file instead of overwriting it")
+	cli.BoolVar(&k, "k", false, "Log key strokes to file as well")
+
+	cli.Parse(os.Args[1:])
+
+	shell, ok := os.LookupEnv("SHELL")
+	if cli.Arg(1) == "" && (!ok || shell == "") {
+		return &Termcording{}, errors.New("shell not set")
+	}
+	interactive := true
+	switch cli.NArg() {
+	case 0:
+		fName = defaultFileName
+		cmdName = shell
+	case 1:
+		fName = cli.Arg(0)
+		cmdName = shell
+	default:
+		fName = cli.Arg(0)
+		cmdName = cli.Arg(1)
+		cmdArgs = cli.Args()[2:]
+		interactive = false
+	}
+
+	return NewTermcording(&Config{
+		Filename:      fName,
+		CmdName:       cmdName,
+		CmdArgs:       cmdArgs,
+		QuietMode:     q,
+		Append:        a,
+		Interactive:   interactive,
+		LogKeystrokes: k,
+		PrintHelp:     h,
+	}, options...)
+}
+
+//Start runs the tc.cmd in a pseudo-terminal and writes all output to tc.out
+func (tc *Termcording) Start() error {
+	if tc.Config.PrintHelp {
+		if tc.out == nil {
+			tc.out = os.Stdout
+		}
+		printHelp(tc)
+		return nil
+	}
+
+	if tc.cmd == nil {
+		tc.cmd = exec.Command(tc.Config.CmdName, tc.Config.CmdArgs...)
+	}
+
+	if !tc.Config.QuietMode {
+		fmt.Println("Starting recording session. CTRL-D to end.")
+		defer fmt.Printf("\nRecording session ended. Session saved to %s\n", tc.Config.Filename)
+	}
+
+	ptmx, restoreMode, err := ptmxFromCmd(tc.cmd, tc.Config.Interactive) //TODO convert to method.
+	if err != nil {
+		return err
+	}
+	defer func() {
+		restoreMode()
+		ptmx.Close()
+	}()
+
+	in := io.Writer(ptmx)
+	if tc.out == nil {
+		mode := os.O_TRUNC
+		if tc.Config.Append {
+			mode = os.O_APPEND
+		}
+		f, err := os.OpenFile(tc.Config.Filename, os.O_WRONLY|os.O_CREATE|mode, 0755)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		if tc.Config.LogKeystrokes {
+			in = io.MultiWriter(ptmx, f)
+		}
+		tc.out = io.MultiWriter(os.Stdout, f)
+	}
+
+	go func() {
+		io.Copy(in, os.Stdin)
+	}()
+
+	io.Copy(tc.out, ptmx)
+
+	return tc.cmd.Wait()
+}
+
+func ptmxFromCmd(c *exec.Cmd, interactive bool) (*os.File, func(), error) {
+	modeRestoreFn := func() {}
+
+	ptmx, err := pty.Start(c)
+	if err != nil {
+		return nil, nil, err
+	}
+	if interactive {
+		// Handle pty size.
+		ch := make(chan os.Signal, 1)
+		signal.Notify(ch, syscall.SIGWINCH)
+		go func() {
+			for range ch {
+				if err := pty.InheritSize(os.Stdin, ptmx); err != nil {
+					log.Printf("error resizing pty: %s", err)
+				}
+			}
+		}()
+		ch <- syscall.SIGWINCH // Initial resize.
```

Comment:
I'm not absolutely clear why this is necessary—if we're sending ourselves a signal to set the terminal size, why not just set it directly?

Reply:
Well this is just for the setting the initial size by taking advantage of that the handler go routine we just started. This is needed because:

1. `pty.Start` does not set size.
2. I hadn't dug deeper into the `pty` library _yet_ and pretty much just copied the basic example. 

In my defence, this was definitely on my refactoring to-do list 😅  Anyway, I am still looking through the library but for I what I can see so far, I could probably use `pty.StartWithSize` and `pty.GetSize` to create the pseudo-terminal instead of just `pty.Start`. If so, we shouldn't need to manually pass in SIGWINCH to the channel just to set the initial size.